File: lib/src/main/java/net/dean/jraw/models/MultiredditPatch.java
package net.dean.jraw.models;

import com.google.auto.value.AutoValue;
import com.squareup.moshi.Json;
import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.Moshi;
import net.dean.jraw.models.internal.SubredditElement;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

@AutoValue
public abstract class MultiredditPatch {
    @Nullable
    @Json(name = "description_md") public abstract String getDescription();

    @Nullable
    @Json(name = "display_name") public abstract String getDisplayName();

    @Nullable
    @Json(name = "icon_name") public abstract String getIconName();

    @Nullable
    @Json(name = "key_color") public abstract String getKeyColor();

    @Nullable
    @Json(name = "subreddits") public abstract List<SubredditElement> getSubreddits();

    @Nullable
    public abstract String getVisibility();

    @Nullable
    public abstract String getWeightingScheme();

    public static final class Builder {
        private String description;
        private String displayName;
        private String iconName;
        private String keyColor;
        private List<String> subreddits;
        private String visibility;
        private String weightingScheme;

        /** Markdown-formatted text */
        public Builder description(String description) { this.description = description; return this; }

        /** The name this multireddit will go by */
        public Builder displayName(String displayName) { this.displayName = displayName; return this; }

        /**
         * According to the API, one of:
         * <ul>
         *   <li>{@code art and design}
         *   <li>{@code ask}
         *   <li>{@code books}
         *   <li>{@code business}
         *   <li>{@code cars}
         *   <li>{@code comics}
         *   <li>{@code cute animals}
         *   <li>{@code diy}
         *   <li>{@code entertainment}
         *   <li>{@code food and drink}
         *   <li>{@code funny}
         *   <li>{@code games}
         *   <li>{@code grooming}
         *   <li>{@code health}
         *   <li>{@code life advice}
         *   <li>{@code military}
         *   <li>{@code models pinup}
         *   <li>{@code music}
         *   <li>{@code news}
         *   <li>{@code philosophy}
         *   <li>{@code pictures and gifs}
         *   <li>{@code science}
         *   <li>{@code shopping}
         *   <li>{@code sports}
         *   <li>{@code style}
         *   <li>{@code tech}
         *   <li>{@code travel}
         *   <li>{@code unusual stories}
         *   <li>{@code video}
         *   <li>{@code (empty string)}
         * </ul>
         */
        public Builder iconName(String iconName) { this.iconName = iconName; return this; }

        /**
         * A hex-formatted hex string, like `#CEE3F8`. This color is primarily used when viewing the multireddit on the
         * mobile site.
         */
        public Builder keyColor(String keyColor) { this.keyColor = keyColor; return this; }

        /** A list of subreddits to include in this multireddit. Do not include the `/r/` prefix. */
        public Builder subreddits(List<String> subreddits) { this.subreddits = subreddits; return this; }
        public Builder subreddits(String... subreddits) { return subreddits(Arrays.asList(subreddits)); }

        /** One of `public`, `private`, or `hidden` */
        public Builder visibility(String visibility) { this.visibility = visibility; return this; }

        /** Either `classic` or `fresh` */
        public Builder weightingScheme(String weightingScheme) { this.weightingScheme = weightingScheme; return this; }

        public MultiredditPatch build() {
            List<SubredditElement> subreddits = new ArrayList<>(this.subreddits == null ? 0 : this.subreddits.size());
            if (this.subreddits != null)
                for (String name : this.subreddits)
                    subreddits.add(SubredditElement.create(name));
            return new AutoValue_MultiredditPatch(description, displayName, iconName, keyColor, subreddits, visibility, weightingScheme);
        }
    }

    public static JsonAdapter<MultiredditPatch> jsonAdapter(Moshi moshi) {
        return new AutoValue_MultiredditPatch.MoshiJsonAdapter(moshi);
    }
}

File: lib/src/main/kotlin/net/dean/jraw/models/MultiredditPatch.kt
package net.dean.jraw.models

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty

@JsonInclude(JsonInclude.Include.NON_NULL)
class MultiredditPatch private constructor(b: Builder) {
    @JsonProperty("description_md")
    val description: String? = b.description
    val displayName: String? = b.displayName
    val iconName: String? = b.iconName
    val keyColor: String? = b.keyColor
    val subreddits: List<SubredditElement> = b.subreddits.map { SubredditElement(it) }
    val visibility: String? = b.visibility
    val weightingScheme: String? = b.weightingScheme

    class Builder {
        internal var description: String? = null
        internal var displayName: String? = null
        internal var iconName: String? = null
        internal var keyColor: String? = null
        internal var subreddits: List<String> = emptyList()
        internal var visibility: String? = null
        internal var weightingScheme: String? = null

        /** Markdown-formatted text */
        fun description(description: String): Builder { this.description = description; return this }

        /** The name this multireddit will go by */
        fun displayName(displayName: String): Builder { this.displayName = displayName; return this }

        /**
         * According to the API, one of :
         *  - `art and design`
         *  - `ask`
         *  - `books`
         *  - `business`
         *  - `cars`
         *  - `comics`
         *  - `cute animals`
         *  - `diy`
         *  - `entertainment`
         *  - `food and drink`
         *  - `funny`
         *  - `games`
         *  - `grooming`
         *  - `health`
         *  - `life advice`
         *  - `military`
         *  - `models pinup`
         *  - `music`
         *  - `news`
         *  - `philosophy`
         *  - `pictures and gifs`
         *  - `science`
         *  - `shopping`
         *  - `sports`
         *  - `style`
         *  - `tech`
         *  - `travel`
         *  - `unusual stories`
         *  - `video`
         *  - `` *(empty string)*
         *  - `None`
         */
        fun iconName(iconName: String): Builder { this.iconName = iconName; return this }

        /**
         * A hex-formatted hex string, like `#CEE3F8`. This color is primarily used when viewing the multireddit on the
         * mobile site.
         */
        fun keyColor(keyColor: String): Builder { this.keyColor = keyColor; return this }

        /** A list of subreddits to include in this multireddit. Do not include the `/r/` prefix. */
        fun subreddits(subreddits: List<String>): Builder { this.subreddits = subreddits; return this }

        /** One of `public`, `private`, or `hidden` */
        fun visibility(visibility: String): Builder { this.visibility = visibility; return this }

        /** Either `classic` or `fresh` */
        fun weightingScheme(weightingScheme: String): Builder { this.weightingScheme = weightingScheme; return this }

        fun build() = MultiredditPatch(this)
    }

    class SubredditElement(val name: String)
}
